wurst/_handles/Camera.wurst

```diff
@@ -0,0 +1,166 @@
+package Camera
+import NoWurst
+import public Vectors
+
+public function getCameraBoundMin() returns vec2
+    return vec2(GetCameraBoundMinX(), GetCameraBoundMinY())
+
+public function getCameraBoundMax() returns vec2
+    return vec2(GetCameraBoundMaxX(), GetCameraBoundMaxY())
+
+public function getCameraTargetPositionVec2() returns vec2
+    return vec2(GetCameraTargetPositionX(), GetCameraTargetPositionY())
+
+public function getCameraTargetPositionVec3() returns vec3
+    let targetPosition = getCameraTargetPositionVec2()
+    return vec3(targetPosition.x, targetPosition.y, GetCameraTargetPositionZ())
+
+public function getCameraEyePositionVec2() returns vec2
+    return vec2(GetCameraEyePositionX(), GetCameraEyePositionY())
+
+public function getCameraEyePositionVec3() returns vec3
+    let eyePosition = getCameraEyePositionVec2()
+    return vec3(eyePosition.x, eyePosition.y, GetCameraEyePositionZ())
+
+public function setCameraPosition(vec2 pos)
+    SetCameraPosition(pos.x, pos.y)
+
+public function setCameraQuickPosition(vec2 pos)
+    SetCameraQuickPosition(pos.x, pos.y)
+
+public function setCameraBounds(vec2 corner1, vec2 corner2, vec2 corner3, vec2 corner4)
+    SetCameraBounds(corner1.x, corner1.y, corner2.x, corner2.y, corner3.x, corner3.y, corner4.x, corner4.y)
+
+public function panCameraTo(vec2 pos)
+    PanCameraTo(pos.x, pos.y)
+
+public function panCameraToTimed(vec2 pos, real duration)
+    PanCameraToTimed(pos.x, pos.y, duration)
+
+public function panCameraToWithZ(vec2 pos, real zOffsetDest)
```

Comment:
why not overload `panCameraTo` with vec3?

Reply:
I actually thought about it, but if I understand the function correctly, this does not pan to a specific point in 3D space. It pans to a 2D space and raises or lowers the camera by the amount or zooms the camera in or out (not sure which, but I assume this is why it's real zOffsetDest). The original function signature is: "native PanCameraToTimedWithZ takes real x, real y, real zOffsetDest, real duration returns nothing"

So here, I am worried that using vec3 might give the wrong impression.

Reply:
It does pan to Z, meaning if there would be duration, then camera movement will be smooth. Yes, Z offset will stay and yes, it is not absolute Z, because wc3's camera is always relative to terrain height. But it's still a certain position, which can be translated to absolute 3D coords by adding terrain height to Z.
But in wurst, vec3 is used in two ways: in absolute 3D space and relatively to terrain height. So the vec3 for your function will be the second type, and so there won't be any wrong impressions.

And even if I'm not right, you can forestall that impression by a hotdoc describing that feature.

Reply:
@Trokkin I'm a bit confused by your explanation here. The description for one of these types of functions in GUI reads: "Pan camera to (Location) with height x.xx above the terrain over y.yy seconds. The camera will not drop below the terrain height during its camera path."

What would I need to do to achieve the same thing in Camera.wurst?

i.e., if I wrote:

```
public function panCameraTo(vec3 pos)
    PanCameraToWithZ(pos.x, pos.y, pos.z)
```

I'm not sure this would produce the correct result. What would I need to do here?

Reply:
actually you could see in what JASS code that GUI functions are translated (they're actually BJ functions from Blizzard.j, btw) and look for anything suspicious.
As I know, the result would be correct, if you mean pos.z as the height *above terrain*, which is still a correct usage for vec3 as I explained.

Reply:
Unfortunately, that doesn't exactly help here. Here is the JASS code:

```
function PanCameraToTimedLocWithZForPlayer takes player whichPlayer, location loc, real zOffset, real duration returns nothing
    if (GetLocalPlayer() == whichPlayer) then
        // Use only local code (no net traffic) within this block to avoid desyncs.
        call PanCameraToTimedWithZ(GetLocationX(loc), GetLocationY(loc), zOffset, duration)
    endif
endfunction
```

And PanCameraToTimedWithZ is a native, so I can't see what it does.

Now if you tell me that vec3.z component automatically defaults to "height above terrain", then it should be fine. Otherwise, I'm not sure how to force vec3 to use terrain height instead of absolute.

Reply:
So, the description should apply 1:1 to that native, as it describes a simple wrapper without any hidden logic.

I told you, that's uncertain: vec3.z is used in both ways as relative and as absolute height, depending on function. See `package Unit` and how vec3 is used there, if you're curious or to find out how to convert back and forth absolute to relative heights. Exactly you want to see getTerrainZ() usage.

Reply:
Can I ask you guys to make sure you document this in lep/jassdoc?